src/java.base/share/classes/java/lang/invoke/MethodHandleProxies.java

```diff
@@ -133,27 +158,32 @@ private MethodHandleProxies() { }  // do not instantiate
      * @throws WrongMethodTypeException if the target cannot
      *         be converted to the type required by the requested interface
      */
-    // Other notes to implementors:
-    // <p>
-    // No stable mapping is promised between the single-method interface and
-    // the implementation class C.  Over time, several implementation
-    // classes might be used for the same type.
-    // <p>
-    // If the implementation is able
-    // to prove that a wrapper of the required type
-    // has already been created for a given
-    // method handle, or for another method handle with the
-    // same behavior, the implementation may return that wrapper in place of
-    // a new wrapper.
-    // <p>
-    // This method is designed to apply to common use cases
-    // where a single method handle must interoperate with
-    // an interface that implements a function-like
-    // API.  Additional variations, such as single-abstract-method classes with
-    // private constructors, or interfaces with multiple but related
-    // entry points, must be covered by hand-written or automatically
-    // generated adapter classes.
-    //
+    /*
+     * Implementor notes:
+     * This API is free to reuse the same implementation instance or class for
+     * suitable inputs. It is not guaranteed to return a new instance for every
+     * call. No stable mapping is promised between the single-method interface
+     * and the wrapper's implementation class. Over time, several implementation
+     * classes might be used for the same type. If the implementation is able to
+     * prove that a wrapper of the required type has already been created for a
+     * given method handle, the implementation may return that wrapper in place
+     * of a new wrapper.
+     *
+     * Discussion:
+     * Since project leyden aims to improve startup speed, asInterfaceInstance
+     * will share one implementation class for each interface than one implementation
+     * class for each method handle. This is good for use cases where multiple distinct
+     * method handles are converted into one interface stored in a collection, where each
+     * instance is invoked rarely (or only once), such as property getters on a bean.
+     *
+     * Using super-customized implementation classes for each method handle would
+     * allow constant-folding of calls through the returned instance, but it comes with
+     * a huge class definition overhead for each instance and is not feasible for rare
+     * invocation use cases like above.
+     *
+     * The shared-class implementation is also closer in behavior to the original
+     * proxy-backed implementation. We might add another API for super-customized instances.
```

Comment:
The implementor note mainly specifies that there is no guarantee on a stable mapping of the SAM interface to the implementation class.

I see this new note you added is to document the current implementation and alternatives.   I would move this closer to the code (see below).   I made some suggested edits.   I avoid using the term "super-customized" since it's not clear what it means unless reading JBS comments.  I also avoid referring to Project Leyden but instead describes it with some reasonable details say "more friendly for pre-generation....".